### DISCUSSION TITLE
Curriocity about weird type inference

### DISCUSSION THREAD

**Discussion by martinkozle** [microsoft/pyright#6275] [General]
Why does adding a tuple `isinstance` check, which I would guess should have no effect here, change the inferred type so much?

Code sample 1 in [pyright playground](https://pyright-play.net/?code=CYUwZgBGAUCGCUAuAUBNECWlYQLy4gCIBnAFwCcBGAJgGZCV0mJyRSBXcgOwllXRAAbLJmIYuZWFwDGIOABpMXUkn7MWbTjz5NWHbhAByAey4hkFgEZ4o0WvGSsAbiFiCA%2BqQCeABzmWHIA)

```python
def f(a):
    if a == "str123":
        return a
    elif isinstance(a, int):
        return a
    return None


b = f(3)
reveal_type(b)

```

Output:

```
Type of "b" is "int | Literal['str123'] | None"
```

Code sample 2 in [pyright playground](https://pyright-play.net/?code=CYUwZgBGAUCGCUAuAUBNECWlYQLy4gCIBnAFwCcBGAJgGZCV0mJyRSBXcgOwllXRAAbLJmIYuZWFwDGIOABpMXUkn7MWbTjz5MhIjGImkpshRA4AHQSFXq0rDt15qNjngDkA9lxDI-AIzwoaFp4ZFYANxBYQQB9UgBPCzl-MKA)

```python
def f(a):
    if a == "str123":
        return a
    elif isinstance(a, int):
        return a
    elif isinstance(a, tuple):
        return a
    return None


b = f(3)
reveal_type(b)

```

Output:

```
Type of "b" is "Literal[3] | None"
```

**Comment by martinkozle** [comment]
To simplify the case a little bit:

Code sample 3 in [pyright playground](https://pyright-play.net/?code=CYUwZgBGAUCGCUAuAUBNECWkMGcMDscAXWfAYxDgBpN8ilV0mAnEIgV2fwlkbVY5cIAOQD2%2BEMikAjCAF4o0AMzxkrAG4hYAGwD6RAJ4AHStNVA)

```python
def f(a):
    if isinstance(a, int):
        return a
    return None


b = f(3)
reveal_type(b)

```

Output:

```
Type of "b" is "int | None"
```

Code sample 4 in [pyright playground](https://pyright-play.net/?code=CYUwZgBGAUCGCUAuAUBNECWkMGcMDscAXWfAYxDgBpN8ilV0mAnEIgV2fwlkbRAA2WTHkIlylWDQ4AHASAZMWbTt17KOXCADkA9vhDIjAIwgBeKNADM8ZKwBuIWAID6RAJ4zKx20A)

```python
def f(a):
    if isinstance(a, int):
        return a
    elif isinstance(a, tuple):
        return a
    return None


b = f(3)
reveal_type(b)

```

Output:

```
Type of "b" is "Literal[3] | None"
```

Why does adding an `isinstance` tuple check narrow the type of the other branch?

**Reply by erictraut** [reply]
In your first example, pyright is able to infer the return type unambiguously. It's always `int | None`. In the second example, pyright first attempts to use its normal return type inference, which produces the answer `int | tuple[Unknown, ...] | None`. If the initial attempt produces a result with an `Unknown`, pyright then tries to use a more sophisticated technique called [call-site return type inferences](https://microsoft.github.io/pyright/#/type-inference?id=call-site-return-type-inference). This technique is limited to small and simple functions with unannotated parameters. It effectively re-analyzes the entire function using parameter types that are derived from the argument types. In your example, you're passing a value of `3`, so it analyzes the function as though parameter `a` has type `Literal[3]`. This produces the result `Literal[3] | None`.

**Reply by martinkozle** [reply]
I didn't know that pyright first tries one algorithm and then falls back to another if the first inference isn't complete. Thank you for the explanation. The behavior makes sense now!
Although, now I have a follow up question of why the second method isn't always used if it narrows the type more accurately?
